services/gundeck/src/Gundeck/ThreadBudget.hs

```diff
@@ -0,0 +1,212 @@
+-- | Like "Brig.Budget", but in-memory, per host (not per service), and with an strict/exact
+-- upper bound.  Like https://hackage.haskell.org/package/token-bucket, but takes the entire
+-- run-time of the actions into account, not just the number of executions.
+-- http://hackage.haskell.org/package/rate-limit also looks related.
+--
+-- USE CASE: keep a lid of stalled native push notification threads.  if SNS is up, there
+-- will be many short-running executions of the action.  when SNS is down, the threads will
+-- accumulate in memory and choke the gundeck instances.  so we want to stop spawning more
+-- threads (and discard or queue native push notifications) before we run out of memory (which
+-- could cause system outages).
+--
+-- http-client connection pools should handle this naturally and without doing anything, but
+-- instead connection pools grow infinitely until system resources (file handles, memory) are
+-- exhausted.  See https://github.com/snoyberg/http-client/issues/307#issuecomment-343829351
+module Gundeck.ThreadBudget
+  ( ThreadBudgetState
+  , mkThreadBudgetState
+  , runWithBudget
+  , watchThreadBudgetState
+
+  -- * for testing
+  , threadBudgetLimits
+  , runningThreads
+  , cancelAllThreads
+  ) where
+
+import Imports
+
+import Control.Exception (ErrorCall(ErrorCall))
+import Control.Exception.Safe (catchAny)
+import Control.Lens
+import Control.Monad.Catch (MonadCatch, throwM)
+import Data.Metrics (Metrics)
+import Data.Metrics.Middleware (gaugeSet, path)
+import Data.SizedHashMap (SizedHashMap)
+import Data.Time
+import Data.UUID (UUID, toText)
+import Data.UUID.V4 (nextRandom)
+import Gundeck.Options
+import UnliftIO.Async
+import UnliftIO.Exception (finally)
+
+import qualified Data.Set as Set
+import qualified Data.SizedHashMap as SHM
+import qualified System.Logger.Class as LC
+
+
+data ThreadBudgetState = ThreadBudgetState
+  { _threadBudgetLimits  :: MaxConcurrentNativePushes
+  , _threadBudgetRunning :: IORef BudgetMap
+  } deriving (Generic)
+
+-- | Store all handles for cleanup in 'watchThreadBudgetState'.
+type BudgetMap = SizedHashMap UUID (Maybe (Async ()))
+
+-- generated with @makeLenses@, but we only want one of the two lenses that generates.
+threadBudgetLimits ::
+  Lens' ThreadBudgetState MaxConcurrentNativePushes
+threadBudgetLimits f_a1A3A (ThreadBudgetState x1_a1A3B x2_a1A3C)
+  = (fmap (\ y1_a1A3D -> (ThreadBudgetState y1_a1A3D) x2_a1A3C))
+      (f_a1A3A x1_a1A3B)
```

Comment:
i'd rather not, since this is generated code (as indicated in the comment).  is it ok if i don't?

Reply:
Why is it important to not have the second lens from `ThreadBudgetState` , i.e. for `threadBudgetRunning` ? (Please add a comment why that would be harmful)

It's no different from hand-writing your Lens though, is it?

But once it's in the source code, it's no longer "generated" automatically in any way, right? So then, nicer naming sounds like it could make the code less ugly to look at.
I.e. I don't understand how to replace `...` in the sentence `this code has to remain with ugly variables, because...`.

Reply:
heh, i just noticed something: there must not be a lens for the limits, since we don't want to allow changing it during runtime of a thread budget state.  it should be a getter, or just a function, which means that i should probably just write it by hand.

this doesn't resolve our bike shedding issue about making generated code prettier by rewriting it, but it's evidence that we're not bike shedding too much, since you put the finger on a real issue.  thanks!  :-)